sal/tests/server/unit/test_data.py

```diff
@@ -0,0 +1,437 @@
+
+    """
+    Tests that a DataTree get operation fails if the path revision is invalid  
+
+    GIVEN
+        A request with a URL path
+        AND a revision query where the revision=negative int
+        AND a request with an object query of full, summary or no object query
+    WHEN
+        The DataTree gets from the path
+    THEN
+        A BadRequest exception is raised
+    """
+
+    path = 'this/is/the/path'
+    data = {'revision':revision}
+    if obj:
+        data['object'] = obj
+
+    _check_raising_request(server, 'get', path, BadRequest, data=data)
+
+
+def test_data_tree_get_invalid_object_query(server):
+
+    """
+    Tests that an exception is raised when the query argument for object is not
+    'full' or 'summary'
+
+    GIVEN
+        A request with a URL path
+        AND a request with a query with revision=int
+        AND a request with a query with an invalid object (not full or summary) 
+    WHEN
+        The DataTree gets from the path
+    THEN
+        A BadRequest exception is raised  
+    """
+
+    path = 'this/is/the/path'
+    data = {'revision':0, 'object':'invalid'}
+    
+    _check_raising_request(server, 'get', path, BadRequest, data=data)
+
+
+@pytest.mark.parametrize('json_content, cls',
+                         [({'content':'object',
+                            'type':'branch',
+                            'object':{'description':'A branch'}},
+                            Branch),
+                          ({'content':'object',
+                            'type':'leaf',
+                            'object':{'description':{'type':'string',
+                                                     'value':'A data object'},
+                                      '_type':{'type':'string',
+                                                'value':'object'},
+                                      '_class':{'type':'string',
+                                                'value':'scalar'},
+                                      '_group':{'type':'string',
+                                                'value':'core'},
+                                      '_version':{'type':'uint64',
+                                                  'value':1},
+                                      'value':{'type':'uint64',
+                                               'value':5}}},
+                            DataObject)])
+def test_data_tree_post_object(server, mock_persistence_provider,
+                               json_content, cls):
+
+    """
+    Tests that a DataTree correctly puts a branch or data object to a path
+    using a PersistenceProvider 
+
+    GIVEN
+        A request with a URL path
+        AND a request with JSON content describing a branch or data object
+    WHEN
+        The DataTree posts to the path
+    THEN
+        The Datatree puts the object onto the PersistenceProvider at the path
+        and returns 204
+    """
+
+    path = 'this/is/the/path'
```

Comment:
**question:** The [docs say](https://simple-access-layer.github.io/documentation/server/rest_api.html#put-leaf-node): "if creating a new [leaf] node, the parent node in the data tree must be a branch node otherwise the request will fail." This test writes to a path with multiple _new_ branch nodes, so are the docs essentially only saying "you may not have a leaf node whose parent is also a leaf node"?

Reply:
**suggestion:** Further to this, a test for attempting to create a leaf node under another leaf node may be useful.

Reply:
So I agree that parents must be branch nodes.  I don't think it allows for adding multiple branches at once though (although I think you could add a branch has a non-empty `branches` which refers to currently existing branches within SAL).

Are you sure that the test writes multiple new nodes?

Reply:
I think I've conflated the URL path with the node that is actually being written, in which case this test looks fine.